File: appIos/feed-test-fixtures/feed-test-fixtures/FeedRobot.swift
import XCTest

/*
 Robot to make assertions on the Feed screen.
 */
public class FeedRobot {
    private let app: XCUIApplication
    
    public init(app: XCUIApplication) {
        self.app = app
    }
    
    public func assertLoadingExists() {
        XCTAssert(app.activityIndicators[Loading].exists)
    }
    
    public func assertLoadingDoesNotExist() {
        XCTAssert(!app.activityIndicators[Loading].exists)
    }
    
    public func assertEmptyStateExists() {
        XCTAssert(app.staticTexts["No results found. Please\ntry another search term"].exists)
    }
    
    public func assertEmptyStateDoesNotExist() {
        XCTAssert(!app.staticTexts["No results found. Please\ntry another search term"].exists)
    }
    
    public func assertSectionExists(header: String) {
        XCTAssert(app.staticTexts[header].exists)
    }
    
    public func assertSectionDoesNotExist(header: String) {
        XCTAssert(!app.staticTexts[header].exists)
    }
    
    public func assertFeedItemExists(
        title: String,
        date: String
    ) {
        XCTAssert(app.staticTexts[title].exists)
        XCTAssert(app.staticTexts[date].exists)
    }
    
    public func assertFeedItemDoesNotExist(
        title: String,
        date: String
    ) {
        XCTAssert(!app.staticTexts[title].exists)
        XCTAssert(!app.staticTexts[date].exists)
    }
    
    public func assertFavouriteItemsShown(favouriteItems: Int) {
        XCTAssertEqual(app.images.matching(identifier: AddedToFavourites).count, favouriteItems)
    }
    
    public func assertBackButtonExists() {
        XCTAssert(app.buttons["Newsfeed"].exists)
    }
    
    public func assertBackButtonDoesNotExist() {
        XCTAssert(!app.buttons["Newsfeed"].exists)
    }
    
    public func assertBlockingErrorExists() {
        XCTAssert(app.staticTexts[BlockingErrorTitle].exists)
        XCTAssert(app.staticTexts[BlockingErrorSubtitle].exists)
        XCTAssert(app.buttons[BlockingErrorAction].exists)
    }
    
    public func assertBlockingErrorDoesNotExist() {
        XCTAssert(!app.staticTexts[BlockingErrorTitle].exists)
        XCTAssert(!app.staticTexts[BlockingErrorSubtitle].exists)
        XCTAssert(!app.buttons[BlockingErrorAction].exists)
    }
    
    public func assertNonBlockingErrorExists() {
        XCTAssert(app.staticTexts[NonBlockingErrorTitle].exists)
        XCTAssert(app.staticTexts[NonBlockingErrorSubtitle].exists)
        XCTAssert(app.buttons[NonBlockingErrorAction].exists)
    }
    
    public func assertNonBlockingErrorDoesNotExist() {
        XCTAssert(!app.staticTexts[NonBlockingErrorTitle].exists)
        XCTAssert(!app.staticTexts[NonBlockingErrorSubtitle].exists)
        XCTAssert(!app.buttons[NonBlockingErrorAction].exists)
    }
    
    public func clickPost(title: String) {
        app.staticTexts[title].tap()
    }
}

private let Loading = "Loading"

private let AddedToFavourites = "Added to favourites"

private let BlockingErrorTitle = "Oups!"
private let BlockingErrorSubtitle = "Something has gone wrong. Please try again."
private let BlockingErrorAction = "Try again"
private let NonBlockingErrorTitle = "Oups"
private let NonBlockingErrorSubtitle = "Something went wrong. Please try again."
private let NonBlockingErrorAction = "Retry"

File: android/feed/test-fixtures/src/main/java/com/gchristov/newsfeed/android/feed/testfixtures/FeedRobot.kt
package com.gchristov.newsfeed.android.feed.testfixtures

import androidx.compose.ui.test.SemanticsNodeInteractionsProvider
import androidx.compose.ui.test.assertCountEquals
import androidx.compose.ui.test.onAllNodesWithContentDescription
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import com.gchristov.newsfeed.commoncomposetest.onIndeterminateProgress

fun SemanticsNodeInteractionsProvider.feed(block: FeedRobot.() -> Unit) {
    FeedRobot(this).block()
}

class FeedRobot internal constructor(private val provider: SemanticsNodeInteractionsProvider) {
    fun assertLoadingExists() {
        provider.onIndeterminateProgress().assertExists()
    }

    fun assertLoadingDoesNotExist() {
        provider.onIndeterminateProgress().assertDoesNotExist()
    }

    fun assertEmptyStateExists() {
        provider.onNodeWithText("No results found. Please\ntry another search term").assertExists()
    }

    fun assertEmptyStateDoesNotExist() {
        provider.onNodeWithText("No results found. Please\ntry another search term")
            .assertDoesNotExist()
    }

    fun assertSectionExists(header: String) {
        provider.onNodeWithText(header).assertExists()
    }

    fun assertSectionDoesNotExist(header: String) {
        provider.onNodeWithText(header).assertDoesNotExist()
    }

    fun assertFeedItemExists(
        title: String,
        date: String,
    ) {
        provider.onNodeWithText(title).assertExists()
        provider.onNodeWithText(date).assertExists()
    }

    fun assertFeedItemDoesNotExist(
        title: String,
        date: String,
    ) {
        provider.onNodeWithText(title).assertDoesNotExist()
        provider.onNodeWithText(date).assertDoesNotExist()
    }

    fun assertFavouriteItemsShown(favouriteItems: Int) {
        provider
            .onAllNodesWithContentDescription(AddedToFavourites)
            .assertCountEquals(favouriteItems)
    }

    fun assertBlockingErrorExists() {
        provider.onNodeWithText(BlockingErrorTitle).assertExists()
        provider.onNodeWithText(BlockingErrorSubtitle).assertExists()
        provider.onNodeWithText(BlockingErrorAction).assertExists()
    }

    fun assertBlockingErrorDoesNotExist() {
        provider.onNodeWithText(BlockingErrorTitle).assertDoesNotExist()
        provider.onNodeWithText(BlockingErrorSubtitle).assertDoesNotExist()
        provider.onNodeWithText(BlockingErrorAction).assertDoesNotExist()
    }

    fun assertNonBlockingErrorExists() {
        provider.onNodeWithText(NonBlockingErrorTitle).assertExists()
        provider.onNodeWithText(NonBlockingErrorSubtitle).assertExists()
        provider.onNodeWithText(NonBlockingErrorAction).assertExists()
    }

    fun assertNonBlockingErrorDoesNotExist() {
        provider.onNodeWithText(NonBlockingErrorTitle).assertDoesNotExist()
        provider.onNodeWithText(NonBlockingErrorSubtitle).assertDoesNotExist()
        provider.onNodeWithText(NonBlockingErrorAction).assertDoesNotExist()
    }

    fun clickPost(title: String) {
        provider.onNodeWithText(title).performClick()
    }
}

private const val AddedToFavourites = "Added to favourites"

private const val BlockingErrorTitle = "Oups!"
private const val BlockingErrorSubtitle = "Something has gone wrong. Please try again."
private const val BlockingErrorAction = "Try again"
private const val NonBlockingErrorTitle = "Oups"
private const val NonBlockingErrorSubtitle = "Something went wrong. Please try again."
private const val NonBlockingErrorAction = "Retry"